File: backend/app/api/endpoints/showcases.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import delete, func, select
from uuid_utils import uuid7

from app.api.deps import CurrentUser, DbSession
from app.models.image import Image
from app.models.showcase import Showcase, ShowcaseImage
from app.schemas.common import MessageResponse
from app.schemas.showcase import (
    ShowcaseCreate,
    ShowcaseDetailResponse,
    ShowcaseImageAdd,
    ShowcaseImageCheck,
    ShowcaseImageCheckResult,
    ShowcaseImageInfo,
    ShowcaseImageRemove,
    ShowcaseImageReorder,
    ShowcaseResponse,
    ShowcaseUpdate,
)

router = APIRouter(prefix="/showcases", tags=["showcases"])


@router.get("", response_model=list[ShowcaseResponse])
async def list_showcases(
    db: DbSession,
    _: CurrentUser,
) -> list[ShowcaseResponse]:
    """List all showcases with image counts."""
    # Get showcases with image counts
    count_subquery = (
        select(
            ShowcaseImage.showcase_id,
            func.count(ShowcaseImage.image_id).label("image_count"),
        )
        .group_by(ShowcaseImage.showcase_id)
        .subquery()
    )

    result = await db.execute(
        select(
            Showcase,
            func.coalesce(count_subquery.c.image_count, 0).label("image_count"),
            Image.thumbnail_path,
        )
        .outerjoin(count_subquery, Showcase.id == count_subquery.c.showcase_id)
        .outerjoin(Image, Showcase.cover_image_id == Image.id)
        .order_by(Showcase.created_at.desc())
    )

    showcases = []
    for row in result:
        showcase = row[0]
        image_count = row[1]
        cover_thumbnail_path = row[2]
        showcases.append(
            ShowcaseResponse(
                id=showcase.id,
                name=showcase.name,
                description=showcase.description,
                icon=showcase.icon,
                cover_image_id=showcase.cover_image_id,
                cover_thumbnail_path=cover_thumbnail_path,
                image_count=image_count,
                created_at=showcase.created_at,
                updated_at=showcase.updated_at,
            )
        )

    return showcases


@router.post("/check-images", response_model=list[ShowcaseImageCheckResult])
async def check_images_in_showcases(
    db: DbSession,
    _: CurrentUser,
    data: ShowcaseImageCheck,
) -> list[ShowcaseImageCheckResult]:
    """Check which images are already in each showcase."""
    # Get count of matching images per showcase
    result = await db.execute(
        select(
            ShowcaseImage.showcase_id,
            func.count(ShowcaseImage.image_id).label("existing_count"),
        )
        .where(ShowcaseImage.image_id.in_(data.image_ids))
        .group_by(ShowcaseImage.showcase_id)
    )

    return [
        ShowcaseImageCheckResult(
            showcase_id=row[0],
            existing_count=row[1],
        )
        for row in result
    ]


@router.post("", response_model=ShowcaseResponse, status_code=status.HTTP_201_CREATED)
async def create_showcase(
    db: DbSession,
    _: CurrentUser,
    data: ShowcaseCreate,
) -> ShowcaseResponse:
    """Create a new showcase."""
    showcase = Showcase(
        id=uuid7(),
        name=data.name,
        description=data.description,
        icon=data.icon,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )
    db.add(showcase)
    await db.commit()
    await db.refresh(showcase)

    return ShowcaseResponse(
        id=showcase.id,
        name=showcase.name,
        description=showcase.description,
        icon=showcase.icon,
        cover_image_id=showcase.cover_image_id,
        cover_thumbnail_path=None,
        image_count=0,
        created_at=showcase.created_at,
        updated_at=showcase.updated_at,
    )


@router.get("/{showcase_id}", response_model=ShowcaseDetailResponse)
async def get_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
) -> ShowcaseDetailResponse:
    """Get a showcase by ID with its images."""
    result = await db.execute(
        select(Showcase, Image.thumbnail_path)
        .outerjoin(Image, Showcase.cover_image_id == Image.id)
        .where(Showcase.id == showcase_id)
    )
    row = result.one_or_none()

    if not row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    showcase = row[0]
    cover_thumbnail_path = row[1]

    # Get images in this showcase
    images_result = await db.execute(
        select(
            Image.id,
            Image.storage_path,
            Image.thumbnail_path,
            ShowcaseImage.sort_order,
            ShowcaseImage.added_at,
        )
        .join(ShowcaseImage, Image.id == ShowcaseImage.image_id)
        .where(ShowcaseImage.showcase_id == showcase_id)
        .where(Image.deleted_at.is_(None))
        .order_by(ShowcaseImage.sort_order)
    )

    images = [
        ShowcaseImageInfo(
            id=row[0],
            storage_path=row[1],
            thumbnail_path=row[2],
            sort_order=row[3],
            added_at=row[4],
        )
        for row in images_result
    ]

    return ShowcaseDetailResponse(
        id=showcase.id,
        name=showcase.name,
        description=showcase.description,
        icon=showcase.icon,
        cover_image_id=showcase.cover_image_id,
        cover_thumbnail_path=cover_thumbnail_path,
        image_count=len(images),
        created_at=showcase.created_at,
        updated_at=showcase.updated_at,
        images=images,
    )


@router.put("/{showcase_id}", response_model=ShowcaseResponse)
async def update_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
    data: ShowcaseUpdate,
) -> ShowcaseResponse:
    """Update an existing showcase."""
    result = await db.execute(select(Showcase).where(Showcase.id == showcase_id))
    showcase = result.scalar_one_or_none()

    if not showcase:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    if data.name is not None:
        showcase.name = data.name
    if data.description is not None:
        showcase.description = data.description if data.description else None
    if data.icon is not None:
        showcase.icon = data.icon if data.icon else None
    if data.cover_image_id is not None:
        showcase.cover_image_id = data.cover_image_id

    showcase.updated_at = datetime.utcnow()

    await db.commit()
    await db.refresh(showcase)

    # Get image count and cover thumbnail path
    count_result = await db.execute(
        select(func.count(ShowcaseImage.image_id)).where(
            ShowcaseImage.showcase_id == showcase_id
        )
    )
    image_count = count_result.scalar() or 0

    cover_thumbnail_path = None
    if showcase.cover_image_id:
        cover_result = await db.execute(
            select(Image.thumbnail_path).where(Image.id == showcase.cover_image_id)
        )
        cover_thumbnail_path = cover_result.scalar_one_or_none()

    return ShowcaseResponse(
        id=showcase.id,
        name=showcase.name,
        description=showcase.description,
        icon=showcase.icon,
        cover_image_id=showcase.cover_image_id,
        cover_thumbnail_path=cover_thumbnail_path,
        image_count=image_count,
        created_at=showcase.created_at,
        updated_at=showcase.updated_at,
    )


@router.delete("/{showcase_id}", response_model=MessageResponse)
async def delete_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
) -> MessageResponse:
    """Delete a showcase."""
    result = await db.execute(select(Showcase).where(Showcase.id == showcase_id))
    showcase = result.scalar_one_or_none()

    if not showcase:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    await db.delete(showcase)
    await db.commit()
    return MessageResponse(message="Showcase deleted successfully")


@router.post("/{showcase_id}/images", response_model=MessageResponse)
async def add_images_to_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
    data: ShowcaseImageAdd,
) -> MessageResponse:
    """Add images to a showcase."""
    # Verify showcase exists
    result = await db.execute(select(Showcase).where(Showcase.id == showcase_id))
    showcase = result.scalar_one_or_none()

    if not showcase:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    # Get max sort_order
    max_order_result = await db.execute(
        select(func.max(ShowcaseImage.sort_order)).where(
            ShowcaseImage.showcase_id == showcase_id
        )
    )
    max_order = max_order_result.scalar() or 0

    # Get existing image IDs in this showcase
    existing_result = await db.execute(
        select(ShowcaseImage.image_id).where(ShowcaseImage.showcase_id == showcase_id)
    )
    existing_ids = {row[0] for row in existing_result}

    # Add new images
    added_count = 0
    for image_id in data.image_ids:
        if image_id not in existing_ids:
            max_order += 1
            showcase_image = ShowcaseImage(
                showcase_id=showcase_id,
                image_id=image_id,
                sort_order=max_order,
                added_at=datetime.utcnow(),
            )
            db.add(showcase_image)
            added_count += 1

    # Update showcase's updated_at
    showcase.updated_at = datetime.utcnow()

    # Set cover image if not set
    if showcase.cover_image_id is None and data.image_ids:
        showcase.cover_image_id = data.image_ids[0]

    await db.commit()
    return MessageResponse(message=f"Added {added_count} images to showcase")


@router.delete("/{showcase_id}/images", response_model=MessageResponse)
async def remove_images_from_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
    data: ShowcaseImageRemove,
) -> MessageResponse:
    """Remove images from a showcase."""
    # Verify showcase exists
    result = await db.execute(select(Showcase).where(Showcase.id == showcase_id))
    showcase = result.scalar_one_or_none()

    if not showcase:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    # Delete the associations
    delete_result = await db.execute(
        delete(ShowcaseImage).where(
            ShowcaseImage.showcase_id == showcase_id,
            ShowcaseImage.image_id.in_(data.image_ids),
        )
    )

    # If cover image was removed, clear it
    if showcase.cover_image_id in data.image_ids:
        showcase.cover_image_id = None

    # Update showcase's updated_at
    showcase.updated_at = datetime.utcnow()

    await db.commit()
    return MessageResponse(
        message=f"Removed {delete_result.rowcount} images from showcase"
    )


@router.put("/{showcase_id}/images/reorder", response_model=MessageResponse)
async def reorder_images_in_showcase(
    db: DbSession,
    _: CurrentUser,
    showcase_id: UUID,
    data: ShowcaseImageReorder,
) -> MessageResponse:
    """Reorder images in a showcase."""
    # Verify showcase exists
    result = await db.execute(select(Showcase).where(Showcase.id == showcase_id))
    showcase = result.scalar_one_or_none()

    if not showcase:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Showcase not found",
        )

    # Update sort_order for each image
    for i, image_id in enumerate(data.image_ids):
        await db.execute(
            ShowcaseImage.__table__.update()
            .where(ShowcaseImage.showcase_id == showcase_id)
            .where(ShowcaseImage.image_id == image_id)
            .values(sort_order=i)
        )

    # Update showcase's updated_at
    showcase.updated_at = datetime.utcnow()

    await db.commit()
    return MessageResponse(message="Images reordered successfully")

File: backend-rs/src/http/showcases.rs
//! Showcase HTTP handlers (mirror endpoints/showcases.py).

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::Json;
use uuid::Uuid;

use super::auth::CurrentUser;
use super::AppState;
use crate::dto::common::{validate_name, MessageResponse};
use crate::dto::showcase::{
    ShowcaseCreate, ShowcaseDetailResponse, ShowcaseImageCheckResult, ShowcaseImageIds,
    ShowcaseResponse, ShowcaseUpdate,
};
use crate::error::AppError;
use crate::showcase;

/// Mirror Field(min_length=1[, max_length=max]) on an image-id list.
fn validate_image_ids(ids: &[Uuid], max: Option<usize>) -> Result<(), AppError> {
    if ids.is_empty() {
        return Err(AppError::BadRequest(
            "image_ids must contain at least 1 item".to_string(),
        ));
    }
    if let Some(max) = max
        && ids.len() > max
    {
        return Err(AppError::BadRequest(format!(
            "image_ids must contain at most {max} items"
        )));
    }
    Ok(())
}

/// GET /api/showcases
pub async fn list_showcases(
    _user: CurrentUser,
    State(state): State<AppState>,
) -> Result<Json<Vec<ShowcaseResponse>>, AppError> {
    Ok(Json(showcase::list(&state.pool).await?))
}

/// POST /api/showcases/check-images
pub async fn check_images(
    _user: CurrentUser,
    State(state): State<AppState>,
    Json(body): Json<ShowcaseImageIds>,
) -> Result<Json<Vec<ShowcaseImageCheckResult>>, AppError> {
    validate_image_ids(&body.image_ids, Some(100))?;
    Ok(Json(
        showcase::check_images(&state.pool, &body.image_ids).await?,
    ))
}

/// POST /api/showcases
pub async fn create_showcase(
    _user: CurrentUser,
    State(state): State<AppState>,
    Json(body): Json<ShowcaseCreate>,
) -> Result<(StatusCode, Json<ShowcaseResponse>), AppError> {
    validate_name(&body.name).map_err(AppError::BadRequest)?;
    let resp = showcase::create(
        &state.pool,
        &body.name,
        body.description.as_deref(),
        body.icon.as_deref(),
    )
    .await?;
    Ok((StatusCode::CREATED, Json(resp)))
}

/// GET /api/showcases/{id}
pub async fn get_showcase(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<ShowcaseDetailResponse>, AppError> {
    let resp = showcase::get_detail(&state.pool, id)
        .await?
        .ok_or_else(|| AppError::NotFound("Showcase not found".to_string()))?;
    Ok(Json(resp))
}

/// PUT /api/showcases/{id}
pub async fn update_showcase(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<ShowcaseUpdate>,
) -> Result<Json<ShowcaseResponse>, AppError> {
    if let Some(name) = &body.name {
        validate_name(name).map_err(AppError::BadRequest)?;
    }
    let resp = showcase::update(
        &state.pool,
        id,
        body.name.as_deref(),
        body.description.as_deref(),
        body.icon.as_deref(),
        body.cover_image_id,
    )
    .await?
    .ok_or_else(|| AppError::NotFound("Showcase not found".to_string()))?;
    Ok(Json(resp))
}

/// DELETE /api/showcases/{id}
pub async fn delete_showcase(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<MessageResponse>, AppError> {
    if !showcase::delete(&state.pool, id).await? {
        return Err(AppError::NotFound("Showcase not found".to_string()));
    }
    Ok(Json(MessageResponse::new("Showcase deleted successfully")))
}

/// POST /api/showcases/{id}/images
pub async fn add_images(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<ShowcaseImageIds>,
) -> Result<Json<MessageResponse>, AppError> {
    validate_image_ids(&body.image_ids, Some(100))?;
    let added = showcase::add_images(&state.pool, id, &body.image_ids)
        .await?
        .ok_or_else(|| AppError::NotFound("Showcase not found".to_string()))?;
    Ok(Json(MessageResponse::new(format!(
        "Added {added} images to showcase"
    ))))
}

/// DELETE /api/showcases/{id}/images
pub async fn remove_images(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<ShowcaseImageIds>,
) -> Result<Json<MessageResponse>, AppError> {
    validate_image_ids(&body.image_ids, Some(100))?;
    let removed = showcase::remove_images(&state.pool, id, &body.image_ids)
        .await?
        .ok_or_else(|| AppError::NotFound("Showcase not found".to_string()))?;
    Ok(Json(MessageResponse::new(format!(
        "Removed {removed} images from showcase"
    ))))
}

/// PUT /api/showcases/{id}/images/reorder
pub async fn reorder_images(
    _user: CurrentUser,
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<ShowcaseImageIds>,
) -> Result<Json<MessageResponse>, AppError> {
    validate_image_ids(&body.image_ids, None)?;
    showcase::reorder_images(&state.pool, id, &body.image_ids)
        .await?
        .ok_or_else(|| AppError::NotFound("Showcase not found".to_string()))?;
    Ok(Json(MessageResponse::new("Images reordered successfully")))
}
